### ISSUE TITLE
questions about bins strategy

### ISSUE THREAD

**Issue by xiaohongniua** [hkust-vgd/riconv#3]
Thanks for your excellent work and code! I have a question here.
The origin paper said that **The local space is then uniformly divided into several bins along pm # »** however, the meaning of the code is that local space is uniformly divided into several bins along m(blue dot) to the centre of all dots rather than m(blue dot) to p(red dot), is it an error or you changed the strategy?

**mean_local = tf.reduce_mean(nn_pts, axis=-2, keepdims=True)**
    **_mean_global = tf.reduce_mean(pts, axis=-2, keepdims=True)_**
    _mean_global = tf.expand_dims(mean_global, axis=-2)_

    nn_pts_local_mean = tf.subtract(nn_pts, mean_local, name=tag + 'nn_pts_local_mean') 
    dists_local_mean = tf.norm(nn_pts_local_mean, axis=-1, keepdims=True) # dist to local mean

    vec = mean_local - nn_pts_center
    vec_dist = tf.norm(vec, axis=-1, keepdims =True)
    vec_norm = tf.divide(vec, vec_dist)
    vec_norm = tf.where(tf.is_nan(vec_norm), tf.ones_like(vec_norm) * 0, vec_norm) 

    nn_pts_local_proj = tf.matmul(nn_pts_local, vec_norm, transpose_b=True)
    nn_pts_local_proj_dot = tf.divide(nn_pts_local_proj, dists_local)
    nn_pts_local_proj_dot = tf.where(tf.is_nan(nn_pts_local_proj_dot), tf.ones_like(nn_pts_local_proj_dot) * 0, nn_pts_local_proj_dot)  # check nan

    nn_pts_local_proj_2 = tf.matmul(nn_pts_local_mean, vec_norm, transpose_b=True)
    nn_pts_local_proj_dot_2 = tf.divide(nn_pts_local_proj_2, dists_local_mean)
    nn_pts_local_proj_dot_2 = tf.where(tf.is_nan(nn_pts_local_proj_dot_2), tf.ones_like(nn_pts_local_proj_dot_2) * 0, nn_pts_local_proj_dot_2)  # check nan

    nn_fts = tf.concat([dists_local, dists_local_mean, nn_pts_local_proj_dot, nn_pts_local_proj_dot_2], axis=-1) # d0 d1 a0 a1
    
    # compute indices from nn_pts_local_proj
    _**vec = mean_global - nn_pts_center**_
    vec_dist = tf.norm(vec, axis=-1, keepdims =True)
    vec_norm = tf.divide(vec, vec_dist)
    **nn_pts_local_proj = tf.matmul(nn_pts_local, vec_norm, transpose_b=True)**

**Comment by xiaohongniua** [comment]
compute indices from nn_pts_local_proj
    vec = mean_global - nn_pts_center
    vec_dist = tf.norm(vec, axis=-1, keepdims =True)
    vec_norm = tf.divide(vec, vec_dist)
    nn_pts_local_proj = tf.matmul(nn_pts_local, vec_norm, transpose_b=True)

it seems that the strategy is the same with origin paper after **mean_global-nn_pts_center** changed to **mean_local - nn_pts_center**

**Comment by cszyzhang** [comment]
> ```
> compute indices from nn_pts_local_proj
> vec = mean_global - nn_pts_center
> vec_dist = tf.norm(vec, axis=-1, keepdims =True)
> vec_norm = tf.divide(vec, vec_dist)
> nn_pts_local_proj = tf.matmul(nn_pts_local, vec_norm, transpose_b=True)
> ```
> 
> it seems that the strategy is the same with origin paper after **mean_global-nn_pts_center** changed to **mean_local - nn_pts_center**

Yes. correct. the results of these two convolution direction are the same. You can run in either way.

**Comment by xiaohongniua** [comment]
> > ```
> > compute indices from nn_pts_local_proj
> > vec = mean_global - nn_pts_center
> > vec_dist = tf.norm(vec, axis=-1, keepdims =True)
> > vec_norm = tf.divide(vec, vec_dist)
> > nn_pts_local_proj = tf.matmul(nn_pts_local, vec_norm, transpose_b=True)
> > ```
> > 
> > 
> > it seems that the strategy is the same with origin paper after **mean_global-nn_pts_center** changed to **mean_local - nn_pts_center**
> 
> Yes. correct. the results of these two convolution direction are the same. You can run in either way.

got it! Thanks

**Comment by dvirginz** [comment]
HI @xiaohongniua :)
Have you used this eventually?
I'm asking because the binning technique as it appears in the code is not rotation invariant, while the method in the paper (as you mentioned above) - is.
